include/ipr/interface

```diff
@@ -1526,6 +1530,13 @@ namespace ipr {
    // In-place multiplication-expression -- "a *= b"
    struct Mul_assign : Binary<Category<mul_assign_cat, Classic>> { };
 
+                                   // -- Narrow --
+   // Checked base to derived class conversion
+   struct Narrow : Binary<Category<narrow_cat>, const Expr&, const Class&> {
```

Comment:
The third template argument should be `const Type&`, not `const Class&` -- the target of the narrowing can be a pointer or a reference type, or any other extended type on the platform.

Since this whole work comes out of the _implicit conversion_ work, it might make sense to add a note here that this is never an implicit conversion.   

Note: I know I suggested, then retracted the suggestion, that the implicit conversion sequence should share a common base class....

Reply:
The idea with this approach is that you would apply class transformation on the type of the expression child. Note in the tests that the expression constraint maintains the pointer.

I like `const Type&` as the child better (and was going to go with it in the first place) as you described but I thought it was at least worth trying out this option here.